transparency/boot_entry.go

```diff
@@ -0,0 +1,220 @@
+
+	requirements, err := policy.ParseRequirements(c.BootPolicy)
+	if err != nil {
+		return
+	}
+
+	proof := pb.(*sigsum.ProofBundle)
+
+	claims, err := policy.ParseStatement(proof.Statement)
+	if err != nil {
+		return
+	}
+
+	// Validate the matching between loaded artifact hashes and
+	// the ones included in the proof bundle.
+	if err = b.validateProofHashes(claims); err != nil {
+		return
+	}
+
+	// Validate the matching between the logged claims and the policy requirements.
+	if err = policy.Validate(requirements, claims); err != nil {
+		// The boot bundle is NOT authorized.
+		return
+	}
+
+	// boot-transparency validation passed, boot bundle is authorized.
+	return
+}
+
+// Hash performs data hashing using SHA-256, which is
+// the same algorithm used by boot-transparency library.
+// Returns the computed hash as hex string.
+func Hash(data *[]byte) (hexHash string) {
+	h := sha256.New()
+	h.Write(*data)
+	hash := h.Sum(nil)
+
+	return hex.EncodeToString(hash)
+}
+
+// Validate the matching between loaded artifact hashes and
+// the ones included in the proof bundle.
+// This step is vital to ensure the correspondence between the artifacts
+// loaded in memory during the boot and the claims that will be validated
+// by the boot-transparency policy function.
+func (b *BootEntry) validateProofHashes(s *policy.Statement) (err error) {
+	for _, a := range *b {
+		if err = a.validateProofHash(s); err != nil {
+			return err
+		}
+	}
+
+	return
+}
+
+func (a Artifact) validateProofHash(s *policy.Statement) (err error) {
+	var h artifact.Handler
+	var found bool
+
+	if err = a.hasValidHash(); err != nil {
+		return
+	}
+
+	for _, claimedArtifact := range s.Artifacts {
+		// The claims are referring to a different artifact
+		// category, try with next block of claims in the statement.
+		if a.Category != claimedArtifact.Category {
+			continue
+		}
+
+		if h, err = artifact.GetHandler(a.Category); err != nil {
+			return
+		}
+
+		// boot-transparency expect to parse requirements in JSON format.
+		requirements, _ := json.Marshal(map[string]string{"file_hash": a.Hash})
+
+		r, err := h.ParseRequirements([]byte(requirements))
+		if err != nil {
+			return err
+		}
+
+		c, err := h.ParseClaims([]byte(claimedArtifact.Claims))
+		if err != nil {
+			return err
+		}
+
+		// The validation logic is safe in the sense that error is returned
+		// if a file hash requested by the boot loader is not present in the
+		// statement for a given artifact category.
+		if err = h.Validate(r, c); err != nil {
+			return fmt.Errorf("loaded boot artifacts do not correspond to the proof bundle ones, file hash mismatch")
+		}
+
+		found = true
+		break
+	}
+
+	if !found {
+		return fmt.Errorf("loaded boot artifacts do not correspond to the proof bundle ones, one or more artifacts are not present in the proof bundle")
+	}
+
+	return
+}
+
+func (a Artifact) hasValidHash() (err error) {
+	h, err := hex.DecodeString(a.Hash)
+
+	if err != nil || len(h) != sha256.Size {
```

Comment:
The hash size should come from a constant in boot-transparency rather than being implicit knowledge here.